src/block_range_scanner/common.rs

```diff
@@ -319,9 +326,72 @@ struct LiveStreamingState<N: Network> {
     previous_batch_end: Option<N::BlockResponse>,
 }
 
-/// Assumes that `min_block <= next_start_block <= end`.
-pub(crate) async fn stream_block_range<N: Network>(
-    min_block: BlockNumber,
+#[must_use]
+pub(crate) async fn stream_historical_range<N: Network>(
+    start: BlockNumber,
+    end: BlockNumber,
+    max_block_range: u64,
+    sender: &mpsc::Sender<BlockScannerResult>,
+    provider: &RobustProvider<N>,
+    reorg_handler: &mut ReorgHandler<N>,
+) -> Option<()> {
+    info!("Getting finalized block number");
+    let finalized = match provider.get_block_number_by_id(BlockNumberOrTag::Finalized.into()).await
+    {
+        Ok(block) => block,
+        Err(e) => {
+            error!(error = %e, "Failed to get finalized block");
+            _ = sender.try_stream(e).await;
+            return None;
+        }
+    };
+
+    // no reorg check for finalized blocks
+    let mut batch_start = start;
+    let finalized_batch_end = finalized.min(end);
+    while batch_start < finalized_batch_end {
+        let batch_end = batch_start.saturating_add(max_block_range - 1).min(end);
```

Comment:
I think this may be a bug. If end > finalized the final batch may send unfinalized blocks.

I believe it should be this:

```suggestion
        let batch_end = batch_start.saturating_add(max_block_range - 1).min(finalized_batch_end);
```

Reply:
There is a way to allow the scanner to stream non-finalized blocks in this "phase" as well, for two reasons:
- it would be the last stream in this phase
- a reorg check could be set immediately after in case a non-finalized block was streamed in this phase.

But now that I think about it, it's currently impossible to test this behavior. Better to apply your suggestion for now, as it's simpler to reason about.

Reply:
The more I think about it, the more I'm in favor of your suggestion, simply because it's simpler, and avoids unnecessary edge cases